Fatal: unsupported endpoint reconfiguration from 2 to 1 lanes during flush

log link: http://hpc-master.lab.mtl.com:8080/blue/organizations/jenkins/ucx/detail/ucx/4838/pipeline/

```
[2020-05-07T16:43:34.797Z] [ RUN      ] udx/test_ucp_sockaddr_destroy_ep_on_err.onesided_bidi_cforce/8 <ud_x>
[2020-05-07T16:43:34.797Z] [     INFO ] Testing 65.65.65.12:0
[2020-05-07T16:43:34.797Z] [     INFO ] server listening on 65.65.65.12:56245
[2020-05-07T16:43:35.725Z] [r-vmb-ppc-jenkins:29155:0:29155]       flush.c:51   Fatal: ep 0x3fffa0e90000: unsupported endpoint reconfiguration from 2 to 1 lanes during flush
[2020-05-07T16:43:35.979Z] 
[2020-05-07T16:43:35.979Z] /scrap/jenkins/workspace/ucx-4/contrib/../src/ucp/rma/flush.c: [ ucp_ep_flush_progress() ]
[2020-05-07T16:43:35.979Z]       ...
[2020-05-07T16:43:35.979Z]        46                                           req->send.flush.num_lanes;
[2020-05-07T16:43:35.979Z]        47         if (diff < 0) {
[2020-05-07T16:43:35.979Z]        48             ucs_fatal("ep %p: unsupported endpoint reconfiguration from %d to %d"
[2020-05-07T16:43:35.979Z] ==>    49                       " lanes during flush", ep, req->send.flush.num_lanes,
[2020-05-07T16:43:35.979Z]        50                       ucp_ep_num_lanes(ep));
[2020-05-07T16:43:35.979Z]        51         }
[2020-05-07T16:43:35.979Z]        52 
[2020-05-07T16:43:35.979Z] 
[2020-05-07T16:43:36.234Z] ==== backtrace (tid:  29155) ====
[2020-05-07T16:43:36.234Z]  0 0x000000000005b898 ucs_debug_print_backtrace()  /scrap/jenkins/workspace/ucx-4/contrib/../src/ucs/debug/debug.c:656
[2020-05-07T16:43:36.234Z]  1 0x0000000000047388 ucp_ep_flush_progress()  /scrap/jenkins/workspace/ucx-4/contrib/../src/ucp/rma/flush.c:49
[2020-05-07T16:43:36.234Z]  2 0x0000000000047608 ucp_ep_flush_completion()  /scrap/jenkins/workspace/ucx-4/contrib/../src/ucp/rma/flush.c:236
[2020-05-07T16:43:36.234Z]  3 0x000000000008074c uct_invoke_completion()  /scrap/jenkins/workspace/ucx-4/contrib/../src/uct/base/uct_iface.h:651
[2020-05-07T16:43:36.234Z]  4 0x000000000008074c uct_ud_iface_dispatch_comp()  /scrap/jenkins/workspace/ucx-4/contrib/../src/uct/ib/ud/base/ud_inl.h:222
[2020-05-07T16:43:36.234Z]  5 0x000000000008074c uct_ud_ep_window_release()  /scrap/jenkins/workspace/ucx-4/contrib/../src/uct/ib/ud/base/ud_ep.c:162
[2020-05-07T16:43:36.234Z]  6 0x000000000008074c uct_ud_ep_process_ack()  /scrap/jenkins/workspace/ucx-4/contrib/../src/uct/ib/ud/base/ud_ep.c:545
[2020-05-07T16:43:36.234Z]  7 0x000000000008074c uct_ud_ep_process_rx()  /scrap/jenkins/workspace/ucx-4/contrib/../src/uct/ib/ud/base/ud_ep.c:759
[2020-05-07T16:43:36.234Z]  8 0x000000000008a36c uct_ud_mlx5_iface_poll_rx()  /scrap/jenkins/workspace/ucx-4/contrib/../src/uct/ib/ud/accel/ud_mlx5.c:469
[2020-05-07T16:43:36.234Z]  9 0x000000000008a36c uct_ud_mlx5_iface_progress()  /scrap/jenkins/workspace/ucx-4/contrib/../src/uct/ib/ud/accel/ud_mlx5.c:518
[2020-05-07T16:43:36.234Z] 10 0x0000000000036494 ucs_callbackq_dispatch()  /scrap/jenkins/workspace/ucx-4/contrib/../src/ucs/datastruct/callbackq.h:211
[2020-05-07T16:43:36.234Z] 11 0x0000000000036494 uct_worker_progress()  /scrap/jenkins/workspace/ucx-4/contrib/../src/uct/api/uct.h:2350
[2020-05-07T16:43:36.234Z] 12 0x0000000000036494 ucp_worker_progress()  /scrap/jenkins/workspace/ucx-4/contrib/../src/ucp/core/ucp_worker.c:2004
[2020-05-07T16:43:36.234Z] 13 0x00000000105958bc ucp_test_base::entity::progress()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/ucp/ucp_test.cc:750
[2020-05-07T16:43:36.234Z] 14 0x00000000105959c0 ucp_test::progress()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/ucp/ucp_test.cc:152
[2020-05-07T16:43:36.234Z] 15 0x000000001057bd08 test_ucp_sockaddr::one_sided_disconnect()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:460
[2020-05-07T16:43:36.234Z] 16 0x0000000010564978 test_ucp_sockaddr_destroy_ep_on_err_onesided_bidi_cforce_Test::test_body()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:763
[2020-05-07T16:43:36.234Z] 17 0x00000000101ecb08 ucs::test_base::run()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/test.cc:290
[2020-05-07T16:43:36.234Z] 18 0x00000000101ecb08 ucs::test_base::TestBodyProxy()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/test.cc:316
[2020-05-07T16:43:36.234Z] 19 0x00000000103ab0a0 ucp_test::TestBody()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/ucp/ucp_test.h:171
[2020-05-07T16:43:36.234Z] 20 0x00000000101c8a7c HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest-all.cc:3562
[2020-05-07T16:43:36.234Z] 21 0x00000000101c8a7c testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest-all.cc:3598
[2020-05-07T16:43:36.234Z] 22 0x00000000101b7c0c testing::Test::Run()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest-all.cc:3635
[2020-05-07T16:43:36.234Z] 23 0x00000000101b7d6c testing::TestInfo::Run()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest-all.cc:3812
[2020-05-07T16:43:36.234Z] 24 0x00000000101b7fb8 testing::TestCase::Run()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest-all.cc:3930
[2020-05-07T16:43:36.234Z] 25 0x00000000101bdfb8 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest-all.cc:5808
[2020-05-07T16:43:36.234Z] 26 0x00000000101be404 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest-all.cc:5725
[2020-05-07T16:43:36.234Z] 27 0x00000000101be404 HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest-all.cc:3562
[2020-05-07T16:43:36.234Z] 28 0x00000000101be404 HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest-all.cc:3598
[2020-05-07T16:43:36.234Z] 29 0x00000000101be404 testing::UnitTest::Run()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest-all.cc:5422
[2020-05-07T16:43:36.234Z] 30 0x0000000010142e58 RUN_ALL_TESTS()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/gtest.h:20059
[2020-05-07T16:43:36.234Z] 31 0x0000000010142e58 main()  /scrap/jenkins/workspace/ucx-4/contrib/../test/gtest/common/main.cc:102
[2020-05-07T16:43:36.234Z] 32 0x0000000000025100 generic_start_main.isra.0()  libc-start.c:0
[2020-05-07T16:43:36.234Z] 33 0x00000000000252f4 __libc_start_main()  ???:0
[2020-05-07T16:43:36.234Z] =================================
[2020-05-07T16:43:36.234Z] Sending notification to sergeyo@mellanox.com
[2020-05-07T16:43:38.757Z] [r-vmb-ppc-jenkins:29155:0:29155] Process frozen...
[2020-05-07T18:59:19.694Z] make: *** [test] Terminated
script returned exit code 124
```

--- a/src/ucp/rma/flush.c
+++ b/src/ucp/rma/flush.c
@@ -44,6 +44,8 @@ static void ucp_ep_flush_progress(ucp_request_t *req)
     /* If the number of lanes changed since flush operation was submitted, adjust
      * the number of expected completions */
     if (ucs_unlikely(req->send.flush.num_lanes != num_lanes)) {
+        ucp_trace_req(req, "ep %p: number of lanes changed from %d to %d",
+                      ep, req->send.flush.num_lanes, num_lanes);
         diff                      = num_lanes - req->send.flush.num_lanes;
         req->send.flush.num_lanes = num_lanes;
         if (diff >= 0) {

--- a/src/ucp/rma/flush.c
+++ b/src/ucp/rma/flush.c
@@ -44,8 +44,8 @@ static void ucp_ep_flush_progress(ucp_request_t *req)
     /* If the number of lanes changed since flush operation was submitted, adjust
      * the number of expected completions */
     if (ucs_unlikely(req->send.flush.num_lanes != num_lanes)) {
-        req->send.flush.num_lanes = num_lanes;
         diff                      = num_lanes - req->send.flush.num_lanes;
+        req->send.flush.num_lanes = num_lanes;
         if (diff >= 0) {
             ucp_trace_req(req,
                           "ep %p: adjusting expected flush completion count by %d",

--- a/src/ucp/rma/flush.c
+++ b/src/ucp/rma/flush.c
@@ -32,8 +32,9 @@ static int ucp_ep_flush_is_completed(ucp_request_t *req)
 
 static void ucp_ep_flush_progress(ucp_request_t *req)
 {
-    ucp_ep_h       ep        = req->send.ep;
-    ucp_lane_map_t all_lanes = UCS_MASK(ucp_ep_num_lanes(ep));
+    ucp_ep_h ep              = req->send.ep;
+    unsigned num_lanes       = ucp_ep_num_lanes(ep);
+    ucp_lane_map_t all_lanes = UCS_MASK(num_lanes);
     ucp_ep_flush_state_t *flush_state;
     ucp_lane_index_t lane;
     ucs_status_t status;
@@ -42,19 +43,23 @@ static void ucp_ep_flush_progress(ucp_request_t *req)
 
     /* If the number of lanes changed since flush operation was submitted, adjust
      * the number of expected completions */
-    if (ucs_unlikely(req->send.flush.num_lanes != ucp_ep_num_lanes(ep))) {
-        diff                            = ucp_ep_num_lanes(ep) -
-                                          req->send.flush.num_lanes;
-        if (diff < 0) {
-            ucs_fatal("ep %p: unsupported endpoint reconfiguration from %d to %d"
-                      " lanes during flush", ep, req->send.flush.num_lanes,
-                      ucp_ep_num_lanes(ep));
+    if (ucs_unlikely(req->send.flush.num_lanes != num_lanes)) {
+        req->send.flush.num_lanes = num_lanes;
+        diff                      = num_lanes - req->send.flush.num_lanes;
+        if (diff >= 0) {
+            ucp_trace_req(req,
+                          "ep %p: adjusting expected flush completion count by %d",
+                          ep, diff);
+            req->send.state.uct_comp.count += diff;
+        } else {
+            /* If we have less lanes, it means we are in error flow and
+             * ucp_worker_set_ep_failed() was completed, so we should have
+             * completed the flush on all lanes.
+             */
+            ucs_assertv(req->send.state.uct_comp.count == 0,
+                        "uct_comp.count=%d num_lanes=%d",
+                        req->send.state.uct_comp.count, num_lanes);
         }
-
-        req->send.state.uct_comp.count += diff;
-        req->send.flush.num_lanes       = ucp_ep_num_lanes(ep);
-        ucs_trace_req("flush req %p: adjusting expected completion count by %d",
-                      req, diff);
     }
 
     ucs_trace("ep %p: progress flush req %p, started_lanes 0x%x count %d", ep,

